packages/framer-motion/src/render/types.ts

```diff
@@ -170,14 +37,109 @@ export type VisualElementOptions<Instance, RenderState = any> = {
     reducedMotionConfig?: ReducedMotionConfig
 }
 
-export type CreateVisualElement<Instance> = (
-    Component: string | React.ComponentType<React.PropsWithChildren<unknown>>,
-    options: VisualElementOptions<Instance>
-) => VisualElement<Instance>
-
 /**
  * A generic set of string/number values
  */
 export interface ResolvedValues {
     [key: string]: string | number
 }
+
+export interface VisualElementEventCallbacks {
+    BeforeLayoutMeasure: () => void
+    LayoutMeasure: (layout: Box, prevLayout?: Box) => void
+    LayoutUpdate: (layout: Axis, prevLayout: Axis) => void
+    Update: (latest: ResolvedValues) => void
+    Render: () => void
+    AnimationStart: (definition: AnimationDefinition) => void
+    AnimationComplete: (definition: AnimationDefinition) => void
+    LayoutAnimationStart: () => void
+    LayoutAnimationComplete: () => void
+    SetAxisTarget: () => void
+    Unmount: () => void
+}
+
+export interface LayoutLifecycles {
+    onBeforeLayoutMeasure?(box: Box): void
+
+    onLayoutMeasure?(box: Box, prevBox: Box): void
+
+    /**
+     * @internal
+     */
+    onLayoutAnimationStart?(): void
+
+    /**
+     * @internal
+     */
+    onLayoutAnimationComplete?(): void
+}
+
+export interface AnimationLifecycles {
```

Comment:
Not, sure but is lifecycle one word or two?
```suggestion
export interface AnimationLifeCycles {
```

Reply:
TIL it is two words

Reply:
Ah but in terms of programming it's normally one word https://reactjs.org/docs/state-and-lifecycle.html